### DISCUSSION TITLE
P2G Version 2 - Beta Release

### DISCUSSION THREAD

**Discussion by philosowaffle** [philosowaffle/peloton-to-garmin#77] [Show and tell]
Been working on switching over to generating FIT files instead of TCX since the FIT schema allows us to capture more data.

You can test it out by using this [Docker Compose File](https://github.com/philosowaffle/peloton-to-garmin/blob/v2/docker-compose.yaml).  Sample [configuration.local.json](https://github.com/philosowaffle/peloton-to-garmin/blob/v2/configuration.example.json).  There is also a [windows executable](https://github.com/philosowaffle/peloton-to-garmin/tree/v2/dist) available.

**New Features:**
* Target Cadence - This adds a Workout overlay to the Cadence graph on Garmin indicating what the goal Cadence was for a given point in time.
* Current FTP - Current FTP is captured in the FIT file, however I'm not positive whether Garmin is actually honoring this field or not when generating the power zones chart.
* Resistance - The data is in the FIT file, but Garmin does not seem to be doing anything with it
* Observability - for the nerds out there added Prometheus and Jaeger and Serilog support.
* GPS data
* Logging/Error Handling

**Not Yet Implemented:**
* Target Power - I need to do a workout that has target power so I have some sample data to work with
* Create mac executables for the non-docker folks

**Comment by philosowaffle** [comment]
The [wiki](https://github.com/philosowaffle/peloton-to-garmin/wiki/v2.x.x) is starting to come together. Still several TODO items, but I think I have the basics documented now.

**Comment by dkmcgowan** [comment]
Great work on this btw...  I switched to the Beta docker container for the latest beta version, will delete some older workouts and run them again through this to help you test.  I didn't see a database.json, I was going to move over my old history but delete only some, then extend how many to pull so I could do some testing.  So far I just left it at 1 to make sure it connected and worked.  I was on the win exe before and the main release.  Did the history change when switching to the new format?

**Reply by dkmcgowan** [reply]
based on the configuration.local.json example it looks like I should have a syncHistory.json but on first run it didn't create one

**Reply by dkmcgowan** [reply]
so I deleted my most recent work out from Garmin and then redid the docker folder/image to start from scratch, it's downloading the most recent workout and the log stops at saying the file is prepped for upload, but it doesn't upload...  I have the parameter to upload true in the Garmin section...  so maybe no syncHistory.json since it hasn't tried to upload yet

**Reply by dkmcgowan** [reply]
I redid it with the main branch and it uploads fine to Garmin, I did that with docker and used the env variables like the example...  I also did not see a database.json or syncHistory.json file...  Those must get stored somewhere else?  I see in the latest when I run it again, in your v2, it says it's already downloaded it so not downloading again, but it could be checking the filenames as well

**Reply by philosowaffle** [reply]
First off, thank you so much for testing this out, it really helps me identify gaps in the documentation or other issues I've glossed over.

1. Good thoughts about the sync history DB file, but I've never fully documented how it works, which may be helpful.  Garmin itself prevents duplicate workouts from being uploaded, so the syncHistory in V2 is not used for that case.  Instead, it is used at a much earlier step, where we are fetching the latest workouts from Peloton, by checking for workouts we have already downloaded/converted we can optimize.  So, the two things you need to do are:
    1. If you already have a particular workout uploaded to Garmin, then you would need to delete it first in order to re-upload.  
    1. Secondly, it requires removing the workout from the syncHistory file so that P2G will re-process that workout.
1. Which leads to the problem with the missing syncHistory.json file.  In the Docker container the file does exist and lives at `/app/syncHistory.json`. P2G would have thrown an error if it failed to find this file so we know it does exist.  But I dont see a way for you to access this file, which means no way to remove workouts from it.  **TODO: I need to revisit this to ensure there is an easy way to modify this db file when running from Docker.** Created #94 

If you don't mind, would you please try the following:
1. Delete the local V2 docker container so you can start from scratch
2. Ensure your docker-compose file matches this one:  https://github.com/philosowaffle/peloton-to-garmin/blob/v2/docker-compose.yaml  I made a mistake on the wiki and was referencing the wrong compose file, thank you for helping me catch that
3. For now just use the `configuration.local.json` file for configuration, I have not had a chance to fully test and document how the ENV variables work.
4. Delete your most recent workout from Garmin
5. Then recreate and run the docker container -> see what happens

**Reply by dkmcgowan** [reply]
For the issue, it stops at a line in the log file "Prepped fit file ... for upload" but it does not upload anything

Under the "Garmin" section I have "Upload": true, your default config has it as false, but I changed it.

So it downloads the most recent but does not upload anything to Garmin

I also let it run last night and it did not run again, so the Polling didn't seem to work, I have Polling in the "App" section set to 3600, so "PollingIntervalSeconds": 3600 and "EnablePolling": true

I can scrub my config and send it to you and my log if that helps

As for the json database, why not for Docker create a mapping in volumes for the folder for the database, so the first time it runs if it sees the mapping it will create it, then it's in your docker volume folders and you can edit it.  I tried to map "./app:/app" in volumes and that threw errors and didn't work

**Reply by philosowaffle** [reply]
Yeah, please scrub and send your config, docker-compose file, and logs when you have a chance. I've been using the docker container for several weeks now without issue, so this is likely just a config problem.

I ran into the same issue as you regarding errors when trying to mount the `app` directory which is why I need to investigate this further.

**Reply by dkmcgowan** [reply]
Here are the 3 files...  I removed a good bit from the log, but left the starts of the lines, if you need more let me know, but there was a lot of data there and I wasn't sure what all was in it, but I saw id's, birthdays, last names, etc...


[configs.zip](https://github.com/philosowaffle/peloton-to-garmin/files/6331436/configs.zip)

**Reply by dkmcgowan** [reply]
also, in the working folder I end up with a fit file in the root, and in downloaded there is 1 json file for the download from peloton, and in the upload folder is also a fit file...  the output folder only has the 1 log file I uploaded

**Reply by dkmcgowan** [reply]
also just noticed that the reason polling wouldn't work would be when I run it, it runs once and updates the logs and then stops itself, it doesn't continue to run

**Reply by philosowaffle** [reply]
I think an error is getting thrown from somewhere after the FIT conversion but before the upload step.  And polling is stopped when an error is thrown. But I'm not sure yet why the error is not getting captured in the logs.

**Reply by philosowaffle** [reply]
> also, in the working folder I end up with a fit file in the root, and in downloaded there is 1 json file for the download from peloton, and in the upload folder is also a fit file... the output folder only has the 1 log file I uploaded

This all sounds correct.  The json file is the peloton data, that was consumed and converted to the fit file.  Because your config specified upload to Garmin, a copy was put in the upload folder.  And at this point something broke down.  The next step is the file in the Upload folder would be sent to Garmin. Upon success, everything in the `working` folder would get cleaned up (deleted). And a new polling interval would start.

**Reply by dkmcgowan** [reply]
if you want to add more logging I can pull an updated image and run things again...  I could send you directly the fit/json file if that will help also but don't want to post those open here for everyone, but it seems like it makes it past that step

**Reply by philosowaffle** [reply]
Yup, just published a new image that adds some more error handling.

Two things you can do for this test run:
1.   In your config, modify SyncHistoryDbPath to be `./output/syncHistory.json`. After an initial run, it _should_ get written to the mounted output directory so you can see and edit it on the fly
2. Now that I have the initial log file, you can set the [Log MinimumLevel](https://github.com/philosowaffle/peloton-to-garmin/wiki/Configuration#observability-logs-config) to `Information`, so you don't have to scrub as much next time around.

I should only need the log file.

**Reply by dkmcgowan** [reply]
that worked and moved the syncHistory to the output folder...  Here is your exception

2021-04-18 15:33:40.593 +00:00 [INF] Encoded FIT file /app/working/20_min_90s_Ride_with_Ally_Love.fit
2021-04-18 15:33:40.873 +00:00 [INF] Encoded FIT file /app/working/20_min_New_Tracks_Ride_with_Olivia_Amato.fit
2021-04-18 15:33:40.878 +00:00 [ERR] Uncaught Exception.
System.NullReferenceException: Object reference not set to an instance of an object.
   at Garmin.GarminUploader.UploadViaPython(String[] files) in /build/src/Garmin/GarminUploader.cs:line 67
   at Garmin.GarminUploader.UploadToGarmin() in /build/src/Garmin/GarminUploader.cs:line 50
   at PelotonToGarminConsole.Program.RunAsync(Configuration config) in /build/src/PelotonToGarminConsole/Program.cs:line 125
   at PelotonToGarminConsole.Program.Main(String[] args) in /build/src/PelotonToGarminConsole/Program.cs:line 84

**Reply by philosowaffle** [reply]
Perfect, I see the issue and can reproduce now. Should have a fix shortly.

**Reply by philosowaffle** [reply]
New image with the fix has been pushed.

**Reply by dkmcgowan** [reply]
that seemed to fix it, most recent one uploaded to Garmin...  I still got this exception in the log, I have seen it one other time, but it looks like it's still running 

2021-04-18 16:15:41.269 +00:00 [INF] Encoded FIT file /app/working/20_min_90s_Ride_with_Ally_Love.fit
2021-04-18 16:15:41.551 +00:00 [INF] Encoded FIT file /app/working/20_min_New_Tracks_Ride_with_Olivia_Amato.fit
2021-04-18 16:15:44.586 +00:00 [INF] GUpload: 2021-04-18 16:15:41,763 [INFO] Try to login on GarminConnect...
2021-04-18 16:15:43,655 [INFO] Logged in as davidkmcgowan@gmail.com
2021-04-18 16:15:44,154 [INFO] Uploaded activity 6629753455 : 20_min_New_Tracks_Ride_with_Olivia_Amato.fit
2021-04-18 16:15:44,566 [INFO] Uploaded activity 6629753510 : 20_min_90s_Ride_with_Ally_Love.fit
2021-04-18 16:15:44,566 [INFO] All done.

2021-04-18 16:15:44.590 +00:00 [ERR] Failed to clean up working directory: working
System.IO.IOException: Resource busy
   at System.IO.FileSystem.RemoveDirectoryInternal(DirectoryInfo directory, Boolean recursive, Boolean throwOnTopLevelDirectoryNotFound)
   at System.IO.FileSystem.RemoveDirectory(String fullPath, Boolean recursive)
   at System.IO.Directory.Delete(String path, Boolean recursive)
   at Common.FileHandling.Cleanup(String workingDir) in /build/src/Common/FileHandling.cs:line 37
2021-04-18 16:15:44.605 +00:00 [INF] Sleeping for 3600 seconds...

**Reply by dkmcgowan** [reply]
Also seems like moving syncHistory.json to the output folder should be a good default because I think otherwise when you take the container down and rebuild it with a new image you will lose whatever you had in the internal container space and therefore it would download everything again

**Reply by philosowaffle** [reply]
Yeah, that error is benign and can be ignored for now. I agree, I will change the default to be placed in the output directory.

**Reply by dkmcgowan** [reply]
It's still running and showing output every hour in the log so far, so looks like it's polling...

thanks for the quick fix!

**Reply by dkmcgowan** [reply]
question for you, how does Garmin know you are uploading a duplicate, do you always use the same ID from Peloton?

**Reply by dkmcgowan** [reply]
new exception for you, I updated the config to get the last 4, which is where I like it, picks up if I do a ride, cool down, maybe stretch, weights all in a day...  So when I updated it, it downloaded 2 more that were already in Garmin, so it would have tried for 4, 2 already were in syncHistory so it skipped, got 2 more, tried to upload those, and then I got this error and it crashed

2021-04-18 18:15:46.377 +00:00 [INF] Sleeping for 3600 seconds...
2021-04-18 18:17:28.759 +00:00 [INF] Encoded FIT file /app/working/20_min_Country_Ride_with_Olivia_Amato.fit
2021-04-18 18:17:29.048 +00:00 [INF] Encoded FIT file /app/working/20_min_2000s_Hip_Hop_Ride_with_Hannah_Corbin.fit
2021-04-18 18:17:32.444 +00:00 [INF] GUpload: 2021-04-18 18:17:29,254 [INFO] Try to login on GarminConnect...
2021-04-18 18:17:31,446 [INFO] Logged in as davidkmcgowan@gmail.com
2021-04-18 18:17:31,892 [INFO] Uploaded activity 6630533736 : 20_min_Country_Ride_with_Olivia_Amato.fit
2021-04-18 18:17:32,424 [INFO] Uploaded activity 6630533774 : 20_min_2000s_Hip_Hop_Ride_with_Hannah_Corbin.fit
2021-04-18 18:17:32,425 [INFO] All done.

2021-04-18 18:17:32.447 +00:00 [ERR] Uncaught Exception.
System.InvalidOperationException: Process must exit before requested information can be determined.
   at System.Diagnostics.Process.EnsureState(State state)
   at System.Diagnostics.Process.get_ExitCode()
   at Garmin.GarminUploader.UploadViaPython(String[] files) in /build/src/Garmin/GarminUploader.cs:line 106
   at Garmin.GarminUploader.UploadToGarmin() in /build/src/Garmin/GarminUploader.cs:line 50
   at PelotonToGarminConsole.Program.RunAsync(Configuration config) in /build/src/PelotonToGarminConsole/Program.cs:line 125
   at PelotonToGarminConsole.Program.Main(String[] args) in /build/src/PelotonToGarminConsole/Program.cs:line 84

**Reply by dkmcgowan** [reply]
I got another one for you, I went to catch my wife up, who does a ton of things on the Peloton, so she was behind 109, it got through about 40 of them and threw this exception

2021-04-18 18:59:04.982 +00:00 [ERR] Uncaught Exception.
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at Conversion.FitConverter.GetWorkoutStepsAndLaps(WorkoutSamples workoutSamples, DateTime startTime, Sport sport, SubSport subSport) in /buil$
   at Conversion.FitConverter.Convert(Workout workout, WorkoutSamples workoutSamples, WorkoutSummary workoutSummary) in /build/src/Conversion/Fi$
   at Conversion.Converter`1.Convert(String format) in /build/src/Conversion/IConverter.cs:line 51
   at Conversion.FitConverter.Convert() in /build/src/Conversion/FitConverter.cs:line 27
   at PelotonToGarminConsole.Program.RunAsync(Configuration config) in /build/src/PelotonToGarminConsole/Program.cs:line 119
   at PelotonToGarminConsole.Program.Main(String[] args) in /build/src/PelotonToGarminConsole/Program.cs:line 84

**Reply by philosowaffle** [reply]
For the first exception,  I may have an idea... will take a bit more investigation on my end.

For the second exception, it looks like it is probably failing to convert a particular workout type that I have not handled correctly.  Anyway for you to tell what the workout was this failed on?  In the `working/downloaded` directory I think you'll be able to find the Peloton json file for this workout, you can send that file my way and I should be able to replay locally to try and reproduce.

**Reply by philosowaffle** [reply]
I have created #95  and #96  for tracking these two.

**Reply by dkmcgowan** [reply]
I couldn't tell which workout was the problem one, it downloaded 85 of them and it failed to process about halfway, I figured it would be easier to have you run the same folder and see where it fails, also will let you process and have other examples to help out
[downloads.tar.gz](https://github.com/philosowaffle/peloton-to-garmin/files/6332592/downloads.tar.gz)

**Reply by dkmcgowan** [reply]
I know you said you would think about the other issue, but more info on that one...  I deleted everything for the container, and the lst 4 rides in Garmin, and ran things again and still got the exception.  I then deleted it all again in container and in Garmin and it happened again


System.InvalidOperationException: Process must exit before requested information can be determined.
   at System.Diagnostics.Process.EnsureState(State state)
   at System.Diagnostics.Process.get_ExitCode()
   at Garmin.GarminUploader.UploadViaPython(String[] files) in /build/src/Garmin/GarminUploader.cs:line 106
   at Garmin.GarminUploader.UploadToGarmin() in /build/src/Garmin/GarminUploader.cs:line 50
   at PelotonToGarminConsole.Program.RunAsync(Configuration config) in /build/src/PelotonToGarminConsole/Program.cs:line 125
   at PelotonToGarminConsole.Program.Main(String[] args) in /build/src/PelotonToGarminConsole/Program.cs:line 84

**Reply by dkmcgowan** [reply]
Also noticed that with the old v1 container it was in my timezone, that is the log file and the uploads to Garmin, with v2 it seems to be all in UTC..  I tried setting the following in the compose file, but not sure the image is using tzdata

environment:
      - "TZ=America/New_York"

**Reply by dkmcgowan** [reply]
I tried mapping this in volumes as well and no luck, still UTC

- "/etc/localtime:/etc/localtime:ro"

Maybe something hardcoded or I'm doing it wrong...

I also deleted last 4 and redid things again and this time I didn't get that InvalidOperationException, so that must be something that can happen or not based on something, race or some other condition or something else

**Reply by philosowaffle** [reply]
Have a theory on the `InvalidOperationException` exception. I have pushed a change to address it, docker image is building now and should be available shortly.

Regarding timezones, that is likely an issue with the date I am writing into the FIT file.  Created #97

**Reply by dkmcgowan** [reply]
Also on timezones it's UTC in the log file, less important but if your putting in fixes I would update log file to match timezone too, or maybe that's intentional...

To help cleanup the log file on that directory delete, you could change FileHandling.cs in Common to

DirectoryInfo directory = new DirectoryInfo(workingDir);
 
foreach (FileInfo file in directory.GetFiles()) {
    file.Delete();
}
 
foreach (DirectoryInfo dir in directory.GetDirectories()) {
    dir.Delete(true);
}


Yes, I should just fork and submit a PR but I'm being lazy

**Reply by philosowaffle** [reply]
Just pushed a commit that addresses:

1. `System.IndexOutOfRangeException: Index was outside the bounds of the array.` - should be able to process your wife's workouts again
2. Directory delete error fixed

New image should be built automatically soon. Timezone fix is still todo.

**Reply by dkmcgowan** [reply]
thanks, I will test it today and let you know...  How does Garmin know if the upload is a duplicate?  I saw one happen in testing and I'm wondering if time is part of it, if so I might want on the time fix to run the large batch on the wifes because I might overshoot how many and it could create duplicates if older ones were uploaded with the time not in UTC

**Reply by philosowaffle** [reply]
Unfortunately Garmin is a bit of a walled garden. I'll have to do some experimenting to see if I can figure out how Garmin is deciding uniqueness. I don't think it can be purely based on timestamp or else you'd never be able to dual record an activity with multiple devices and upload them.  But timestamp could be used in conjunction with another field/s to create a unique hash.

**Reply by dkmcgowan** [reply]
that makes sense, I was just curious...  I had one end up in there twice, 4 hours apart, an older one from v1 that had EST and it duplicated with v2 with UTC, so time is probably one part of the key..  Probably title and time hash or something

**Reply by dkmcgowan** [reply]
just tested latest on my account, worked and the directory error was gone...  It wasn't running today and I didn't think I stopped it yesterday, but there were not relevant errors in the log, last entry on the log was the normal Sleeping for XXX seconds message...  I will keep it running tonight and see if it is running tomorrow...  I'm going to wait on the time update before running on the wifes account, so can validate that index out of range fix in combo with that one...  thanks!

**Reply by philosowaffle** [reply]
Latest image should support timezone config now
```
environment:
- TZ=America/New_York
```

Seemed to fix both the log file and the activity timestamps.

**Reply by dkmcgowan** [reply]
That kinda worked, I set the TZ=America/New_York and the output came in EST time, so UTC -5, but it's currently EDT right now, so it should have been UTC -4, so it was still 1 hour off...

**Reply by dkmcgowan** [reply]
I got into the container with docker exec bash and executed "date" and the time is correct, so maybe it's in the parsing of the FIT file that doesn't account for EDT vs EST

**Reply by dkmcgowan** [reply]
also have not seen that invalidoperation exception again and I think I would have by now, so that might be fixed!

**Reply by philosowaffle** [reply]
Hmmmm.....  I'm in central and am correctly getting CDT time for logs/workouts. Prior to this change my workouts/logs were in UTC. But I'm probably still missing something and will have to fiddle with it more.  

The [local timezone offset](https://github.com/philosowaffle/peloton-to-garmin/blob/v2/src/Conversion/FitConverter.cs#L148) is calculated here based on the system Local time.  Although it just now occurred to me this is only applied to the `endTime`.  I need to revisit the FIT specification and see if I can figure out what I'm missing.  Right now they are all UTC with the exception of that one `LocalTimeStamp` message.

**Reply by dkmcgowan** [reply]
things are still running fine since yesterday, and also checked, the time is accurate in the log file, so it's EDT (-4) and worked with my TZ=America/New_York...  It's just the Garmin end result seems to be EST (-5)

**Reply by dkmcgowan** [reply]
still running fine, no issues in the logs!  today my ride automatically uploaded, thanks for all the fixes and quick turnaround!  I still need to verify that parsing fix you did for my wifes data, but I was holding out to see if the EST vs EDT issues had a resolution or not, mainly because I'm worried I won't count and overshoot how many to import and if the time is off it might cause duplicates for things already in Garmin

**Reply by philosowaffle** [reply]
Nice! Glad to hear its running well for someone else.

 The Garmin timezone issue is the next thing I plan to look at, but unfortunately it may be a couple days before I have time.  Just so you know what to expect in terms of turn around.

**Reply by dkmcgowan** [reply]
that's great and no problem at all.... glad I could help get your v2 closer to main, no one wants to support and maintain two versions

**Reply by philosowaffle** [reply]
I had a literal "aha" moment in the shower regarding the timezone+dst and have pushed out another fix. Give it a go and let me know what you see.

**Reply by philosowaffle** [reply]
Also, I want to add, thank you so much for contributing some time to testing this out with me. Your data + your wifes data has allowed me to test over a variety of use cases I might never have covered on my own.  I really appreciate it.

**Reply by dkmcgowan** [reply]
no problem on the testing, glad to do it.  Also, great news, the timezone issue is fixed!  I will run another test with my wifes account tonight and hopefully that works, if so then all my reported issues are fixed, then I will run it for a couple weeks on both our accounts and post back, if all looks good, your that much closer to not supporting v1 anymore

**Reply by dkmcgowan** [reply]
so I ran my wifes and no parsing issues, it downloaded 104 and encoded them but only uploaded 68 to Garmin...  The 68 in the log match what ended up in Garmin, but not sure where the other 36 went and nothing interesting in the logs

**Reply by dkmcgowan** [reply]
so I deleted the history and ran it again, same 104, but left what was already in Garmin there and it picked the same 68 and had the "already uploaded" message in the logs

**Reply by dkmcgowan** [reply]
so it was missing one of the most recent 5 so I ran it again for just 5, not the 100, and it picked up the missing one and uploaded it to Garmin fine, so something was happening when it was trying to do so many

**Reply by dkmcgowan** [reply]
I ran it again with debug and don't see much, there are some no distance lug found and no speed lug found debug messages, but no exceptions

**Reply by philosowaffle** [reply]
Could you send me the upload process portion of the logs from the Debug run? There should be a message that says: `Beginning Garmin Upload.`, copy from there to the end.  The only thing you may want to scrub will be your garmin user name which should only appear once, you can just ctrl-F to find it if you don't want it exposed.

My gut tells me we are running into some kind of limit on the command line.

P2G does all of the conversion logic in dotnet, then uses Process.Start to execute a python command using `gupload` python library to upload the files.  It passes a list of fully qualified file paths to the `gupload` command.

On windows the argument max length is around 32k characters, and in the P2G container the limit is ~100k characters.  We can estimates that `104 workouts x 75characters = 8100` well below either limit.

This is a good find, I'll need to investigate further.

**Reply by dkmcgowan** [reply]
I haven't had a chance yet but will recreate and run the test again for you soon

**Reply by deteuf** [reply]
Hey good job so far
Do you need more device data?
here you have one more for Instinct Solar
<Creator xsi:type="Device_t">
        <Name>Instinct Solar</Name>
        <UnitId>3337746152</UnitId>
        <ProductID>3466</ProductID>
        <Version>
          <VersionMajor>14</VersionMajor>
          <VersionMinor>20</VersionMinor>
          <BuildMajor>0</BuildMajor>
          <BuildMinor>0</BuildMinor>
        </Version>
      </Creator>

**Comment by noiwontsmile** [comment]
I've installed the latest p2g 2 and had initial success with docker but after a few successful uploads I get an authentication error pasted below.  No change in the username/ password between a successful upload and a failed.  

p2g | [14:58:36 INF] Beginning Garmin Upload.
p2g | [14:58:39 INF] GUpload: 2021-06-07 14:58:37,287 [INFO] Try to login on GarminConnect...
p2g | 2021-06-07 14:58:39,346 [CRITICAL] Login Failure: Login check failed.
p2g |
p2g | [14:58:39 ERR] Failed to upload workouts. GUpload had exit code of 1
p2g | [14:58:39 ERR] GUpload returned an error code. Failed to upload workouts.
p2g | Garmin.GarminUploadException: GUpload Exit Code: 1 - GUpload returned an error code. Failed to upload workouts.
p2g | at Garmin.GarminUploader.UploadViaPython(String[] files) in /build/src/Garmin/GarminUploader.cs:line 117
p2g | at Garmin.GarminUploader.UploadToGarmin() in /build/src/Garmin/GarminUploader.cs:line 54
p2g | at PelotonToGarminConsole.Program.RunAsync(Configuration config) in /build/src/PelotonToGarminConsole/Program.cs:line 157
p2g | [14:58:39 WRN] GUpload failed to upload files. You can find the converted files at ./output
p2g | You can manually upload your files to Garmin Connect, or wait for P2G to try again on the next sync job.
p2g | [14:58:39 INF] Sleeping for 86400 seconds...
